### ISSUE TITLE
Per-process Cpu Usage is incorrectly calculated on Apple Silicon

### ISSUE THREAD

**Issue by MillQK** [oshi/oshi#2346]
Hi! I'm testing OSHI 6.4.1 (great library!) on Apple macbook with m2 max cpu. I've faced an issue where processes cpu usage from oshi-demo gui application doesn't match to top / activity monitor app. I've read F.A.Q. and there's an information about difference between windows task manager and OSHI, but nothing about MacOS activity monitor, so I assume that data should be close enough.
I've observed that I need to multiply OSHI's %CPU value to 40 to get needed result, but I can't understand why.

Additional info about cpu from sysctl:
```
sysctl -a | grep machdep.cpu

machdep.cpu.cores_per_package: 12
machdep.cpu.core_count: 12
machdep.cpu.logical_per_package: 12
machdep.cpu.thread_count: 12
machdep.cpu.brand_string: Apple M2 Max
```

The attached files shows the issue.
<img width="1475" alt="Screenshot 2023-04-05 at 15 58 02" src="https://user-images.githubusercontent.com/17243082/230107471-2dbcc62a-647d-459c-b847-4e29a233c55e.png">
<img width="1481" alt="Screenshot 2023-04-05 at 15 57 47" src="https://user-images.githubusercontent.com/17243082/230107487-d995ef26-ee38-4201-b09e-4717263c181c.png">

**Comment by dbwiddis** [comment]
Thanks for the report. Sorry I completely missed this issue until now.  

I was able to reproduce this myself. Initially I was going to say try "of one Processor" vs. System, but that makes it even worse!  There's clearly something amiss in the calculations.

<img width="1677" alt="Screenshot 2023-04-30 at 11 36 33 PM" src="https://user-images.githubusercontent.com/9291703/235417522-8b0d63bf-7df7-4118-9fc3-b09ea7b91ac8.png">

**Comment by dbwiddis** [comment]
OK, here's one obvious bug. 🤦   Per-proc divides by the number of CPUs which makes the percentage smaller; it should be multiplying:
https://github.com/oshi/oshi/blob/c204766fba19c448d38b47e39a7b0dc06b5eda6b/oshi-demo/src/main/java/oshi/demo/gui/ProcessPanel.java#L157-L165

The per-proc interpretation (with the bug fixed) is the correct way to interpret `top` outputs on *nix.

That makes the numbers closer (multiplying the "system" numbers by 12), but it's still too low (as you indicate it looks like it should be 40, meaning we need another 3.3x.

<img width="1677" alt="Screenshot 2023-04-30 at 11 43 54 PM" src="https://user-images.githubusercontent.com/9291703/235418236-6c443aa0-8d3b-4a6f-88a2-82da29104ba1.png">

**Comment by dbwiddis** [comment]
Or, actually, the CPU % shown on `top` is the cumulative value, so the numbers are correct there (other than refresh rate/synchronization issues).

So it's just the division where there should be multiplication... affects all *nix, really. Surprised nobody noticed it until now.

**Comment by dbwiddis** [comment]
Seems [this site](https://blog.guillaume-gomez.fr/articles/2021-09-06+sysinfo%3A+how+to+extract+systems%27+information#Process-CPU-usage) has reported the same behavior:
> On mac, there are two ways to compute a process CPU usage. For some reason, since the new macbook with M1 processors, the "old" one isn't accurate anymore. We'll only describe the "new" one here.

And the "new" one grabs up time from an elapsed tick calculation:
> Before doing anything, we need to get the "system time interval". We get it by retrieving the "ticks" count from `host_processor_info`

Haven't dug into it completely, but it appears that x86 records ticks in milliseconds, while aarch64 has something closer to the Linux "Hz" inmplementation (`_SC_CLK_TCK`).   This [StackOverflow post](https://stackoverflow.com/questions/69318594/how-to-get-the-true-value-of-the-system-clock-in-raw-ticks-i-e-mach-absolute-t) hints at another source:
> On an Intel mac, `mach_absolute_time()` returns the system clock in nanoseconds, and `mach_timebase_info()` returns a 1:1 ratio of nanoseconds to clock ticks.

That `mach_timebase_info` hint is a good search keyword, giving results like [this](https://eclecticlight.co/2020/11/27/inside-m1-macs-time-and-logs/):
> Over the last 14 years, we became accustomed to all Macs being based on Intel processors and chipsets, which encourages bad habits and unrealistic expectations. One of those has been that the internal clock which keeps Mach Absolute Time ticks once every nanosecond. It thus provides a high precision way of measuring time which is also simple to read and manipulate. In the M1 Macs, time has changed, and now ticks every 41.67 nanoseconds, which makes a big difference.

So your estimate of 40 was pretty close.  Looks like you need to multiply your CPU usage by 41.67 on Apple Silicon.  

I'll work on a way to "accurately" calculate this and keep the stats up to date (along with fixing the per-processor-percent bug).

**Comment by dbwiddis** [comment]
Summary for the TL:dr folks:
1. The oshi-demo gui has a bug dividing by number of processors instead of multiplying for per-processor values.
2. When running natively on Apple Silicon (not rosetta) the CPU usage undercounts by a factor of 125/3.  A quick workaround is to multiply usage percentages by this constant until this bug is fixed.  No correction is needed if operating under rosetta, though.

**Comment by dbwiddis** [comment]
Well, that was a bit of an adventure, but I got it figured out!

<img width="1651" alt="Screenshot 2023-05-01 at 10 45 21 PM" src="https://user-images.githubusercontent.com/9291703/235588437-72c9b26e-526c-418a-9e4a-3a6f979bf152.png">

Interestingly, this tick-based counter only impacts the process tick internal array, and doesn't affect the other tick counter sources.  So we just need to apply this correction in one place.

However, while the `mach_timebase_info` ratio would solve the problem on native arm64, most JVMs still run under Rosetta (see the "Intel" line on the gui process above) and that correction is not available there.  Fortunately, now that I knew I was looking for a "timebase" I recalled the frequencies on the cpus on the device tree (stored at boot so robust to rosetta) and the `timebase-frequency` value of 24,000,000 is clock ticks per second. 1000/24 = 125/3 = 41.667.

PR shortly.
